pkg/kubelet/cm/dra/manager.go

```diff
@@ -0,0 +1,246 @@
+					resourceClaim.UID, resourceClaim.Name, resourceClaim.Status.Allocation.ResourceHandle, err)
+			}
+
+			klog.V(3).InfoS("NodePrepareResource succeeded", "response", response)
+
+			annotations, err := generateCDIAnnotations(resourceClaim.UID, driverName, response.CdiDevices)
+			if err != nil {
+				return fmt.Errorf("failed to generate container annotations, err: %+v", err)
+			}
+
+			// Cache prepared resource
+			err = m.cache.add(
+				resourceClaim.Name,
+				resourceClaim.Namespace,
+				&claimInfo{
+					driverName:  driverName,
+					claimUID:    resourceClaim.UID,
+					claimName:   resourceClaim.Name,
+					namespace:   resourceClaim.Namespace,
+					podUIDs:     sets.New(string(pod.UID)),
+					cdiDevices:  response.CdiDevices,
+					annotations: annotations,
+				})
+			if err != nil {
+				return fmt.Errorf(
+					"failed to cache prepared resource, claim: %s(%s), err: %+v",
+					resourceClaim.Name,
+					resourceClaim.UID,
+					err,
+				)
+			}
+		}
+	}
+
+	return nil
+}
+
+// getContainerInfo gets a container info from the claimInfo cache.
+// This information is used by the caller to update a container config.
+func (m *ManagerImpl) getContainerInfo(pod *v1.Pod, container *v1.Container) (*ContainerInfo, error) {
+	annotations := []kubecontainer.Annotation{}
+
+	for i, podResourceClaim := range pod.Spec.ResourceClaims {
+		claimName := resourceclaim.Name(pod, &pod.Spec.ResourceClaims[i])
+
+		for _, claim := range container.Resources.Claims {
+			if podResourceClaim.Name != claim.Name {
+				continue
+			}
+
+			claimInfo := m.cache.get(claimName, pod.Namespace)
+			if claimInfo == nil {
+				return nil, fmt.Errorf("unable to get resource for namespace: %s, claim: %s", pod.Namespace, claimName)
+			}
+
+			klog.V(3).InfoS("add resource annotations", "claim", claimName, "annotations", claimInfo.annotations)
+			annotations = append(annotations, claimInfo.annotations...)
+		}
+	}
+
+	return &ContainerInfo{Annotations: annotations}, nil
+}
+
+// PrepareResources calls plugin NodePrepareResource from the registered DRA resource plugins.
+func (m *ManagerImpl) PrepareResources(pod *v1.Pod, container *v1.Container) (*ContainerInfo, error) {
+	if err := m.prepareContainerResources(pod, container); err != nil {
+		return nil, err
+	}
+
+	return m.getContainerInfo(pod, container)
+}
+
+func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
+	// Call NodeUnprepareResource RPC for every resource claim referenced by the pod
+	for i := range pod.Spec.ResourceClaims {
+		claimName := resourceclaim.Name(pod, &pod.Spec.ResourceClaims[i])
+		claimInfo := m.cache.get(claimName, pod.Namespace)
+
+		// Skip calling NodeUnprepareResource if claim info is not cached
+		if claimInfo == nil {
+			continue
+		}
+
+		// Skip calling NodeUnprepareResource if pod is not cached
+		if !claimInfo.referencedByPod(pod.UID) {
+			continue
```

Comment:
Is this situtation expected / possible? Do we need to log some error here that something went wrong if we call `UnprepareResources` and we can't find a pod using it?

Reply:
This is not an error condition.

All calls to `NodePrepareResource()` and `NodeUnprepareResource()` need to be idempotent, and may end up being triggered multiple times for the same pod. This check then ensures that we don't attempt to **actually** unprepare the resource more than once, but rather assume it has already successfully been done.

Reply:
Although looking at this again, I'm not convinced continuing on this check alone is sufficient here. We should also check that the provided claim info has been removed from the cache (which it wouldn't be if there there was a transient error on the underlying API call, for example).

Reply:
Although looking at this again, we do need someway of ensuring that a transient failure on the underlying `client.NodeUnprepareResource()` call triggers it to be called again. With this check as it is now, this will never happen.

Reply:
Reading through the logic, I think this entire check can simply be removed.

With this removed, we will always run `NodeUnprepareResource()` from this point, so long as a claimInfo object exists and all pod references have been removed from it. Once `NodeUnprepareResource()` returns successfully the claimInfo object is removed, so the check above for `if claimInfo == nil` will prevent us from calling `NodeUnprepareResource()` again.

Reply:
should be fixed by https://github.com/pohly/kubernetes/pull/45

Reply:
Which was merged...